File: python/maximumDifference.py
# Source: https://www.geeksforgeeks.org/maximum-difference-between-two-elements/
# Using minElement and maxDiff
# Time Complexity: O(n)

def maxDiff():
    minEle = array[0]
    maxDiff = array[1] - array[0]

    for i in range(0, len(array)):
        if array[i] < minEle:
            minEle = array[i]

        if array[i] - minEle > maxDiff:
            maxDiff = array[i] - minEle

    return maxDiff

array = [2, 3, 10, 6, 4, 8, 1]
print(maxDiff())

File: java/src/maximumDifference.java
// Source: https://www.geeksforgeeks.org/maximum-difference-between-two-elements/
// using finding maximum subarray concept
// Time Complexity: O(n)

public class maximumDifference {
    public static void main(String[] args) {
        int[] array = {2, 3, 10, 6, 4, 8, 2};
        System.out.println(maxDiff(array));
    }

    private static int maxDiff(int[] array) {
        int diff = 0, maxSoFar = 0, maxEndingHere = 0;

        for(int i = 0; i < array.length - 1; i++) {
            diff = array[i+1] - array[i];

            maxEndingHere += diff;

            if(maxEndingHere < 0) maxEndingHere = 0;
            if(maxSoFar < maxEndingHere) maxSoFar = maxEndingHere;
        }

        return maxSoFar;
    }
}
